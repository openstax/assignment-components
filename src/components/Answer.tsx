import cn from 'classnames';
import { ReactNode } from 'react';
import { ALPHABET, isAnswerChecked, isAnswerCorrect, isAnswerIncorrect } from '../utils';
import { Answer as AnswerType, ChosenAnswer, ID } from '../types';
import { Content } from './Content';
import { SimpleFeedback } from './Feedback';

export interface AnswerProps {
  answer: AnswerType;
  iter: number;
  qid: ID;
  type: 'teacher-review' | 'teacher-preview' | 'student' | 'student-mpp';
  hasCorrectAnswer?: boolean;
  onChangeAnswer: () => void;
  disabled: boolean;
  chosenAnswer: ChosenAnswer;
  correctAnswerId?: ID | null;
  incorrectAnswerId?: ID;
  onKeyPress?: () => void;
  answered_count?: number;
  correctIncorrectIcon?: ReactNode,
  radioBox?: ReactNode;
  contentRenderer?: JSX.Element;
  show_all_feedback: boolean;
}

export const Answer = (props: AnswerProps) => {
  const {
    type,
    iter,
    answer,
    disabled,
    onKeyPress,
    qid,
    chosenAnswer,
    correctAnswerId,
    incorrectAnswerId,
    hasCorrectAnswer,
    answered_count,
    contentRenderer,
    show_all_feedback
  } = props;

  let body, feedback, selectedCount;

  const isChecked = isAnswerChecked(answer, chosenAnswer);
  const isCorrect = isAnswerCorrect(answer, correctAnswerId);
  const isIncorrect = isAnswerIncorrect(answer, incorrectAnswerId);

  const correctIncorrectIcon = (
    <div className="correct-incorrect">
      {isCorrect && props.correctIncorrectIcon}
    </div>
  );

  const classes = cn('answers-answer', {
    'disabled': disabled,
    'answer-checked': isChecked,
    'answer-correct': isCorrect && type !== 'student-mpp',
    'answer-incorrect': incorrectAnswerId && isAnswerIncorrect(answer, incorrectAnswerId),
  });

  let ariaLabel = `${isChecked ? 'Selected ' : ''}Choice ${ALPHABET[iter]}`;
  // somewhat misleading - this means that there is a correct answer,
  // not necessarily that this answer is correct
  if (hasCorrectAnswer) {
    ariaLabel += `(${isCorrect ? 'Correct' : 'Incorrect'} Answer)`;
  }
  ariaLabel += ':';

  let onChangeAnswer, radioBox;

  if (!hasCorrectAnswer
    && (type !== 'teacher-review')
    && (type !== 'teacher-preview')
    && (type !== 'student-mpp')) {
    ({ onChangeAnswer } = props);
  }

  if (onChangeAnswer) {
    radioBox = (
      <input
        type="radio"
        className="answer-input-box"
        checked={isChecked}
        id={`${qid}-option-${iter}`}
        name={`${qid}-options`}
        onChange={onChangeAnswer}
        disabled={disabled}
      />
    );
  }

  if (show_all_feedback && answer.feedback_html) {
    feedback = (
      <SimpleFeedback key="question-mc-feedback" contentRenderer={contentRenderer}>
        {answer.feedback_html}
      </SimpleFeedback>
    );
  }

  if (type === 'teacher-review') {
    let percent = 0;
    if (answer.selected_count && answered_count) {
      percent = Math.round((answer.selected_count / answered_count) * 100);
    }
    selectedCount = (
      <span
        className="selected-count"
        data-percent={`${percent}`}
      >
        {answer.selected_count}
      </span>
    );

    body = (
      <div className="review-wrapper">
        <div className={cn('review-count', { 'green': isCorrect, 'red': !isCorrect })}>
          {selectedCount}
          <span className={cn('letter', { 'green': isCorrect, 'red': !isCorrect })}>
            {ALPHABET[iter]}
          </span>
        </div>

        <div className="answer-answer">
          <Content className="answer-content" component={contentRenderer} html={answer.content_html} />
          {feedback}
        </div>
      </div>
    );
  } else {
    body = (
      <>
        {type === 'teacher-preview' && correctIncorrectIcon}
        {selectedCount}
        {radioBox}
        <label
          onKeyPress={onKeyPress}
          htmlFor={`${qid}-option-${iter}`}
          className="answer-label">
          <span className="answer-letter-wrapper">
            <button
              onClick={onChangeAnswer}
              aria-label={ariaLabel}
              className="answer-letter"
              disabled={disabled || isIncorrect}
              data-test-id={`answer-choice-${ALPHABET[iter]}`}
            >
              {ALPHABET[iter]}
            </button>
          </span>
          <div className="answer-answer">
            <Content className="answer-content" component={contentRenderer} html={answer.content_html} />
            {feedback}
          </div>
        </label>
      </>
    );
  }

  return (
    <div className="openstax-answer">
      <section role="region" className={classes}>
        {body}
      </section>
    </div>
  );
}
Answer.displayName = 'OSAnswer';
