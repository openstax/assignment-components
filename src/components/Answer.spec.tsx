import { Answer, AnswerProps } from './Answer';
import renderer from 'react-test-renderer';
import { answerContent } from '../test/fixtures';

describe('Answer', () => {
  let props: AnswerProps;

  beforeEach(() => {
    props = {
      type: 'student',
      iter: 1,
      answer: {
        id: 1,
        correctness: null,
        isCorrect: true,
        content_html: answerContent[0],
        selected_count: 5
      },
      onChangeAnswer: () => jest.fn(),
      disabled: false,
      onKeyPress: () => jest.fn(),
      qid: 1,
      hasCorrectAnswer: false,
      chosenAnswer: [''],
      correctAnswerId: 2,
      incorrectAnswerId: 0,
      answered_count: 10,
      show_all_feedback: false
    };
  });

  it('matches snapshot', () => {
     const tree = renderer.create(
       <Answer {...props} />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders with a custom renderer if set', () => {
    const CustomRenderer = ({ html = '' }: { html?: string }) => (
      <div className="custom-renderer" dangerouslySetInnerHTML={{ __html: html }} />
    );
    const tree = renderer.create(
      <Answer {...props} contentRenderer={<CustomRenderer />} />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders feedback', () => {
    props.show_all_feedback = true;
    props.answer.feedback_html = '<div>Insightful commentary</div>';
    const tree = renderer.create(
      <Answer {...props} />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders a correct answer', () => {
    const tree = renderer.create(
      <Answer {...props} correctAnswerId={props.answer.id} hasCorrectAnswer={true} disabled={true} />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders an incorrect answer', () => {
    const tree = renderer.create(
      <Answer {...props} incorrectAnswerId={props.answer.id} hasCorrectAnswer={true} />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders a checked answer', () => {
    const tree = renderer.create(
      <Answer {...props} chosenAnswer={[props.answer.id]} />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders teacher review', () => {
    const tree = renderer.create(
      <Answer {...props} type='teacher-review' />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders teacher preview', () => {
    props = {...props, correctAnswerId: props.answer.id, correctIncorrectIcon: <span>Iconic</span>};
    const tree = renderer.create(
      <Answer {...props} type='teacher-preview' />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
