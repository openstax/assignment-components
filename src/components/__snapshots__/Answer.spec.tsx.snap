// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Answer matches snapshot 1`] = `
<div
  className="openstax-answer"
>
  <section
    className="answers-answer"
    role="region"
  >
    <input
      checked={false}
      className="answer-input-box"
      disabled={false}
      id="1-option-1"
      name="1-options"
      onChange={[Function]}
      type="radio"
    />
    <label
      className="answer-label"
      htmlFor="1-option-1"
      onKeyPress={[Function]}
    >
      <span
        className="answer-letter-wrapper"
      >
        <button
          aria-label="Choice b:"
          className="answer-letter"
          data-test-id="answer-choice-b"
          disabled={false}
          onClick={[Function]}
        >
          b
        </button>
      </span>
      <div
        className="answer-answer"
      >
        <span
          className="answer-content"
          dangerouslySetInnerHTML={
            Object {
              "__html": "<ol>
  <li>matter is moving at speeds of less than roughly 1% the speed of light,</li>
  <li>objects are too small to be seen with the naked eye, and</li>
  <li>there is the involvement of only a weak gravitational field</li>
</ol>
",
            }
          }
        />
      </div>
    </label>
  </section>
</div>
`;

exports[`Answer renders a checked answer 1`] = `
<div
  className="openstax-answer"
>
  <section
    className="answers-answer answer-checked"
    role="region"
  >
    <input
      checked={true}
      className="answer-input-box"
      disabled={false}
      id="1-option-1"
      name="1-options"
      onChange={[Function]}
      type="radio"
    />
    <label
      className="answer-label"
      htmlFor="1-option-1"
      onKeyPress={[Function]}
    >
      <span
        className="answer-letter-wrapper"
      >
        <button
          aria-label="Selected Choice b:"
          className="answer-letter"
          data-test-id="answer-choice-b"
          disabled={false}
          onClick={[Function]}
        >
          b
        </button>
      </span>
      <div
        className="answer-answer"
      >
        <span
          className="answer-content"
          dangerouslySetInnerHTML={
            Object {
              "__html": "<ol>
  <li>matter is moving at speeds of less than roughly 1% the speed of light,</li>
  <li>objects are too small to be seen with the naked eye, and</li>
  <li>there is the involvement of only a weak gravitational field</li>
</ol>
",
            }
          }
        />
      </div>
    </label>
  </section>
</div>
`;

exports[`Answer renders a correct answer 1`] = `
<div
  className="openstax-answer"
>
  <section
    className="answers-answer disabled answer-correct"
    role="region"
  >
    <label
      className="answer-label"
      htmlFor="1-option-1"
      onKeyPress={[Function]}
    >
      <span
        className="answer-letter-wrapper"
      >
        <button
          aria-label="Choice b(Correct Answer):"
          className="answer-letter"
          data-test-id="answer-choice-b"
          disabled={true}
        >
          b
        </button>
      </span>
      <div
        className="answer-answer"
      >
        <span
          className="answer-content"
          dangerouslySetInnerHTML={
            Object {
              "__html": "<ol>
  <li>matter is moving at speeds of less than roughly 1% the speed of light,</li>
  <li>objects are too small to be seen with the naked eye, and</li>
  <li>there is the involvement of only a weak gravitational field</li>
</ol>
",
            }
          }
        />
      </div>
    </label>
  </section>
</div>
`;

exports[`Answer renders an incorrect answer 1`] = `
<div
  className="openstax-answer"
>
  <section
    className="answers-answer answer-incorrect"
    role="region"
  >
    <label
      className="answer-label"
      htmlFor="1-option-1"
      onKeyPress={[Function]}
    >
      <span
        className="answer-letter-wrapper"
      >
        <button
          aria-label="Choice b(Incorrect Answer):"
          className="answer-letter"
          data-test-id="answer-choice-b"
          disabled={true}
        >
          b
        </button>
      </span>
      <div
        className="answer-answer"
      >
        <span
          className="answer-content"
          dangerouslySetInnerHTML={
            Object {
              "__html": "<ol>
  <li>matter is moving at speeds of less than roughly 1% the speed of light,</li>
  <li>objects are too small to be seen with the naked eye, and</li>
  <li>there is the involvement of only a weak gravitational field</li>
</ol>
",
            }
          }
        />
      </div>
    </label>
  </section>
</div>
`;

exports[`Answer renders feedback 1`] = `
<div
  className="openstax-answer"
>
  <section
    className="answers-answer"
    role="region"
  >
    <input
      checked={false}
      className="answer-input-box"
      disabled={false}
      id="1-option-1"
      name="1-options"
      onChange={[Function]}
      type="radio"
    />
    <label
      className="answer-label"
      htmlFor="1-option-1"
      onKeyPress={[Function]}
    >
      <span
        className="answer-letter-wrapper"
      >
        <button
          aria-label="Choice b:"
          className="answer-letter"
          data-test-id="answer-choice-b"
          disabled={false}
          onClick={[Function]}
        >
          b
        </button>
      </span>
      <div
        className="answer-answer"
      >
        <span
          className="answer-content"
          dangerouslySetInnerHTML={
            Object {
              "__html": "<ol>
  <li>matter is moving at speeds of less than roughly 1% the speed of light,</li>
  <li>objects are too small to be seen with the naked eye, and</li>
  <li>there is the involvement of only a weak gravitational field</li>
</ol>
",
            }
          }
        />
        <aside>
          <div
            className="question-feedback-content has-html"
            dangerouslySetInnerHTML={
              Object {
                "__html": "<div>Insightful commentary</div>",
              }
            }
          />
        </aside>
      </div>
    </label>
  </section>
</div>
`;

exports[`Answer renders teacher preview 1`] = `
<div
  className="openstax-answer"
>
  <section
    className="answers-answer answer-correct"
    role="region"
  >
    <div
      className="correct-incorrect"
    >
      <span>
        Iconic
      </span>
    </div>
    <label
      className="answer-label"
      htmlFor="1-option-1"
      onKeyPress={[Function]}
    >
      <span
        className="answer-letter-wrapper"
      >
        <button
          aria-label="Choice b:"
          className="answer-letter"
          data-test-id="answer-choice-b"
          disabled={false}
        >
          b
        </button>
      </span>
      <div
        className="answer-answer"
      >
        <span
          className="answer-content"
          dangerouslySetInnerHTML={
            Object {
              "__html": "<ol>
  <li>matter is moving at speeds of less than roughly 1% the speed of light,</li>
  <li>objects are too small to be seen with the naked eye, and</li>
  <li>there is the involvement of only a weak gravitational field</li>
</ol>
",
            }
          }
        />
      </div>
    </label>
  </section>
</div>
`;

exports[`Answer renders teacher review 1`] = `
<div
  className="openstax-answer"
>
  <section
    className="answers-answer"
    role="region"
  >
    <div
      className="review-wrapper"
    >
      <div
        className="review-count red"
      >
        <span
          className="selected-count"
          data-percent="50"
        >
          5
        </span>
        <span
          className="letter red"
        >
          b
        </span>
      </div>
      <div
        className="answer-answer"
      >
        <span
          className="answer-content"
          dangerouslySetInnerHTML={
            Object {
              "__html": "<ol>
  <li>matter is moving at speeds of less than roughly 1% the speed of light,</li>
  <li>objects are too small to be seen with the naked eye, and</li>
  <li>there is the involvement of only a weak gravitational field</li>
</ol>
",
            }
          }
        />
      </div>
    </div>
  </section>
</div>
`;

exports[`Answer renders with a custom renderer if set 1`] = `
<div
  className="openstax-answer"
>
  <section
    className="answers-answer"
    role="region"
  >
    <input
      checked={false}
      className="answer-input-box"
      disabled={false}
      id="1-option-1"
      name="1-options"
      onChange={[Function]}
      type="radio"
    />
    <label
      className="answer-label"
      htmlFor="1-option-1"
      onKeyPress={[Function]}
    >
      <span
        className="answer-letter-wrapper"
      >
        <button
          aria-label="Choice b:"
          className="answer-letter"
          data-test-id="answer-choice-b"
          disabled={false}
          onClick={[Function]}
        >
          b
        </button>
      </span>
      <div
        className="answer-answer"
      >
        <div
          className="custom-renderer"
          dangerouslySetInnerHTML={
            Object {
              "__html": "<ol>
  <li>matter is moving at speeds of less than roughly 1% the speed of light,</li>
  <li>objects are too small to be seen with the naked eye, and</li>
  <li>there is the involvement of only a weak gravitational field</li>
</ol>
",
            }
          }
        />
      </div>
    </label>
  </section>
</div>
`;
